keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -292,18 +293,50 @@ is scoped to.
 or map. These will be accessible to CEL via `self`. The elements of a map or list can be validated using the CEL support for collections
 like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey,
 <predicate>)`.
-  
-- For immutability use case, validator will have access to the existing version of the object. This
-  will be accessible to CEL via the `oldSelf` identifier.
-  - This will only be available on mergable collection types such as objects (unless
-    `x-kubernetes-map-type=atomic`), maps with `x-kubernetes-map-type=granular` and lists
-    with `x-kubernetes-list-type` set to `set` or `map`.  See [Merge
-    Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for
-    details.
-  - The use of "old" is congruent with how `AdmissionReview` identifies the existing object as
-    `oldObject`.
-  - xref [analysis of possible interactions with immutability and
-    validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
+
+- For certain use cases (e.g. immmutability), rules may access to the existing version of the object (see [use cases and other considerations](#access-to-existing-object-state)).
+
+  - Not all update operations are considered comparable for the purposes of this mechanism:
+
+    - Both an old and a new value must exist. It will remain possible to check if a value has been
+      added or removed by placing update rules on the parent node.
+
+    - The path to the schema node being validated must unambiguously refer to the same node -- in
+      terms of identity -- in the new object as it did in the old object. For example, array items
```

Comment:
how disruptive to the current openapi validation code is it to locate/correlate old array items in lists with `x-kubernetes-list-type: set` or `x-kubernetes-list-type: map`? would `oldSelf` be set to the item at the same index in the old object or the item with the same key?

this sort of gets into the same weeds the CRD immutability KEP got into in https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1101-immutable-fields/README.md#openapi-extension-x-kubernetes-immutable-keys

Reply:
For `x-kubernetes-list-type: map`: We can correlate using map key.

`x-kubernetes-list-type: set`: I don't think we need to correlate since ["set applies to lists that include only scalar elements"](https://kubernetes.io/docs/reference/using-api/_print/#merge-strategy) all validation can be declared one the parent (the entire set) and there is no need correlate the individual elements of the set and allow for per element rules.  E.g. if you want to prevent the set to be append only, the rule check that none of the existing elements are removed is defined on the entire set.